dhis-2/dhis-services/dhis-service-validation/src/main/java/org/hisp/dhis/outlierdetection/service/DefaultOutlierDetectionService.java

```diff
@@ -55,119 +46,59 @@
-      error = new ErrorMessage(ErrorCode.E2202);
-    } else if (request.getOrgUnits().isEmpty()) {
-      error = new ErrorMessage(ErrorCode.E2203);
-    } else if (request.getThreshold() <= 0) {
-      error = new ErrorMessage(ErrorCode.E2204);
-    } else if (request.getMaxResults() <= 0) {
-      error = new ErrorMessage(ErrorCode.E2205);
-    } else if (request.getMaxResults() > MAX_LIMIT) {
-      error = new ErrorMessage(ErrorCode.E2206, MAX_LIMIT);
-    } else if (request.hasDataStartEndDate()
-        && request.getDataStartDate().after(request.getDataEndDate())) {
-      error = new ErrorMessage(ErrorCode.E2207);
-    }
+  public OutlierDetectionResponse getOutlierValues(OutlierDetectionRequest request)
+      throws IllegalQueryException {
+    validate(request);
 
-    return error;
+    final OutlierDetectionResponse response = new OutlierDetectionResponse();
+    response.setOutlierValues(getOutliers(request));
+    response.setMetadata(getMetadata(request, response.getOutlierValues()));
+    return response;
   }
 
   @Override
-  public OutlierDetectionRequest getFromQuery(OutlierDetectionQuery query) {
-    OutlierDetectionRequest.Builder request = new OutlierDetectionRequest.Builder();
-
-    List<DataSet> dataSets = idObjectManager.getByUid(DataSet.class, query.getDs());
-
-    // Re-fetch data elements to maintain access control
-
-    List<String> de =
-        dataSets.stream()
-            .map(DataSet::getDataElements)
-            .flatMap(Collection::stream)
-            .filter(d -> d.getValueType().isNumeric())
-            .map(DataElement::getUid)
-            .collect(Collectors.toList());
-
-    de.addAll(query.getDe());
-
-    List<DataElement> dataElements = idObjectManager.getByUid(DataElement.class, de);
-    List<OrganisationUnit> orgUnits =
-        idObjectManager.getByUid(OrganisationUnit.class, query.getOu());
-
-    request
-        .withDataElements(dataElements)
-        .withStartEndDate(query.getStartDate(), query.getEndDate())
-        .withOrgUnits(orgUnits)
-        .withDataStartDate(query.getDataStartDate())
-        .withDataEndDate(query.getDataEndDate());
-
-    if (query.getAlgorithm() != null) {
-      request.withAlgorithm(query.getAlgorithm());
-    }
-
-    if (query.getThreshold() != null) {
-      request.withThreshold(query.getThreshold());
-    }
-
-    if (query.getOrderBy() != null) {
-      request.withOrderBy(query.getOrderBy());
-    }
-
-    if (query.getMaxResults() != null) {
-      request.withMaxResults(query.getMaxResults());
-    }
+  public void getOutlierValuesAsCsv(OutlierDetectionRequest request, Writer writer)
+      throws IllegalQueryException, IOException {
+    JacksonCsvUtils.toCsv(getOutlierValues(request).getOutlierValues(), OutlierValue.class, writer);
+  }
 
-    return request.build();
+  @Override
+  public void getOutlierValuesAsXml(OutlierDetectionRequest request, OutputStream outputStream)
```

Comment:
Is there any specific reason for not supporting all other types of files for the download?

Reply:
ignore

Reply:
Old implementation should not extend the API interface